### PR TITLE
WTrackWidgetGroup with track color background

### PR DESCRIPTION
**Pull request by daschuer**
This is an inherited WWidgetGroup that allows to color the background by the loaded track.

### CODE REVIEW

**Comment by daschuer** [comment]
In Shade, I have colored the background of the default cover and the back ground of the title and artist row:  

grafik

**Comment by ronso0** [comment]
I'll test this the next days and see if / how it can be added to the other skins.

Should we also prepare to change font colors of children labels, or do we stick with low opacity anyway?

**Comment by daschuer** [comment]
I have also thought about changing the font color, but I did not implemented it because I had no specific use case.
If you have one, we can model after it.

**Comment by ronso0** [comment]
I don't have a specifc use case, I was just thinking we might need this when we use the track color with high opacity in skins/schemes with strong colors, like the LateNight Flip theme for example. Then strong track colors will definitely conflict with the skin theme.

I'm more worried about if &/ how we make the track colors optional.
a) If people don't assign colors to tracks they won't care.
b) If people use colors troughout their library they'll probably(!) appreciate having the colors in the decks, too. _Probably_ because I'm still questioning what the actual use of the track color in the decks is. They're great for finding tracks in the library, but in the decks?
c) And there's also a intersecting user group. Me for example, I probably won't use track colors that often but I definitely don't want them to repaint deck sections. I'd prefer a fully saturated color bar somewhere near the cover art for example. And even that not for the information but simply for the comfort of directly selecting a color (future feature) instead of having to go through the track menu (which is awesome btw).

**Comment by daschuer** [comment]
The use case for the track color in the deck is obvious for me. It informs the user which color the track has. The new widget also supports to place a full saturated rectangle somewhere, or just a thin line. 

I have experimented wit this in Shade but did not find a good looking solution this. It looks quite good to just color the default cover art like in the screenshot. But it looks quite ugly if we combine the selected color with the cover art, since that are conflicting colors in most cases.

**Comment by Holzhaus** [comment]
What about coloring the spinny indicator?

**Comment by ronso0** [comment]
okay, I see the reason for the color in th decks.

Spinny indicator might work, but it's a bit hard to see when it's moving. Also spinnies are not always enabled.
Here are some ideas, with **3** being my favourite.

**1** We could put a fully saturated horizontal color bar in between Title | Artist  that shows up only if the track color is set (behind the Artist/Title container so we still have the maximum track drop area):
83340234-ea3fb980-a2d5-11ea-976b-42a50a495975

**2** or even add a colored frame to the cover art:
83340234-ea3fb980-a2d5-11ea-976b-42a50a495975


**3** without colored background:
83340234-ea3fb980-a2d5-11ea-976b-42a50a495975

**Comment by Holzhaus** [comment]
I'd prefer 3 as well, but I'm not a Shade user.

**Comment by daschuer** [comment]
3 looks indeed pealing with the default cover but it fails with many other covers, because you cannot visual distinguish the colored border from the cover which making it look ugly. 

I vote for 3 without the cover art border.

**Comment by Holzhaus** [comment]
> 3 looks indeed pealing with the default cover but it fails with many other covers, because you cannot visual distinguish the colored border from the cover which making it look ugly.
> 
> I vote for 3 without the cover art border.

Sounds good.

**Comment by daschuer** [comment]
grafik

**Comment by Be-ing** [comment]
How about showing WTrackMenu on right click?

**Comment by daschuer** [comment]
Here it is. 

Done.

**Comment by Holzhaus** [comment]
CI fails due to an incomplete type.

**Comment by daschuer** [comment]
Now it works.

**Comment by ronso0** [comment]
> How about showing WTrackMenu on right click?

This is required only if make WTrackWidgetGroup a top-level widget, not stacked underneath.
In that case it should also be a drop target for tracks. In #2736 for example I eliminated 'dead' areas in the artist/title section to improve the UX.

**Comment by daschuer** [comment]
All green. Merge?

File: src/widget/wtrackwidgetgroup.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+    if (m_trackColor.isValid()) {
+        QStylePainter p(this);
+
+        p.fillRect(rect(), QBrush(m_trackColor));
+    }
+}
+
+void WTrackWidgetGroup::mouseMoveEvent(QMouseEvent* event) {
+    if ((event->buttons() & Qt::LeftButton) && m_pCurrentTrack) {
```
Comment (Holzhaus) [inline src/widget/wtrackwidgetgroup.cpp:104]: I think this is a bit more readable:
```suggestion
    if (event->buttons().testFlag(Qt::LeftButton) && m_pCurrentTrack) {
```

File: src/widget/wtrackwidgetgroup.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+#include <QStylePainter>
+#include <QUrl>
+
+#include "control/controlobject.h"
+#include "util/dnd.h"
+#include "widget/wtrackmenu.h"
+
+namespace {
+
+int kDefaultTrackColorAplha = 255;
```
Comment (Holzhaus) [inline src/widget/wtrackwidgetgroup.cpp:14]:
```suggestion
int kDefaultTrackColorAlpha = 255;
```

**Review by Holzhaus — CHANGES_REQUESTED** [review]
Thank for working on this, works fine. Just some minor comments.

**Comment by Holzhaus** [comment]
There's a merge conflict now.

**Comment by daschuer** [comment]
The CI Failure is the known pre-commit issue. 
Merge?

**Comment by Holzhaus** [comment]
There are some unresolved comments above.

File: src/skin/legacyskinparser.cpp [hunk 3]
```diff
@@ -1067,6 +1065,49 @@ QWidget* LegacySkinParser::parseTrackProperty(const QDomElement& node) {
     return p;
 }
 
+QWidget* LegacySkinParser::parseTrackWidgetGroup(const QDomElement& node) {
+    QString channelStr = lookupNodeGroup(node);
+    const char* pSafeChannelStr = safeChannelString(channelStr);
+    BaseTrackPlayer* pPlayer = m_pPlayerManager->getPlayer(channelStr);
+
+    if (!pPlayer) {
```
[the remark this answers is not in the document]
Comment (daschuer) [reply src/skin/legacyskinparser.cpp:1060]: This is not a good Idea, because it happens in case of skin errors. 
Crashing Mixxx in this case is not a good idea. I will give a skin warning instead.

**Comment by daschuer** [comment]
Done

File: src/widget/wtrackwidgetgroup.cpp (hunk 2 above)
Comment (Holzhaus) [reply src/widget/wtrackwidgetgroup.cpp:14]: ping

File: src/widget/wtrackwidgetgroup.cpp (hunk 1 above)
Comment (Holzhaus) [reply src/widget/wtrackwidgetgroup.cpp:105]: ping

File: src/test/cuecontrol_test.cpp [hunk 4]
```diff
@@ -289,12 +289,10 @@ TEST_F(CueControlTest, SeekOnLoadDefault_CueInPreroll) {
...
             ConfigValue(static_cast<int>(SeekOnLoadMode::MainCue)));
     TrackPointer pTrack = createTestTrack();
-    pTrack->setCuePoint(CuePosition(-100.0));
-
     loadTrack(pTrack);
 
-    EXPECT_DOUBLE_EQ(-100.0, m_pCuePoint->get());
-    EXPECT_DOUBLE_EQ(-100.0, getCurrentSample());
+    EXPECT_DOUBLE_EQ(0.0, m_pCuePoint->get());
+    EXPECT_DOUBLE_EQ(0.0, getCurrentSample());
```
Comment (ninomp) [inline src/test/cuecontrol_test.cpp:295]: Unrelated change. This looks like a "leak" from #2800.

**Comment by daschuer** [comment]
Ok, now it should be really ready.

**Comment by ninomp** [comment]
I tried merging this to my local 2.3 branch, but CoverArt is no longer displayed below the library sidebar.

File: .travis.yml [hunk 5]
```diff
@@ -169,7 +169,7 @@ jobs:
         - cd cmake_build
         - cmake -L $CMAKEFLAGS $CMAKEFLAGS_EXTRA -DCMAKE_PREFIX_PATH=${MIXXX_ENVPATH} -DQt5_DIR=${QT_DIR} ..
         - cmake --build .
-        - cpack -G DragNDrop
+        - cpack --verbose -G DragNDrop
```
Comment (Holzhaus) [inline .travis.yml:172]: I don't think this belong into this PR.

File: src/widget/wtrackwidgetgroup.cpp [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+void WTrackWidgetGroup::setup(const QDomNode& node, const SkinContext& context) {
+    WWidgetGroup::setup(node, context);
+
+    bool ok = false;
+    int trackColorAlpha = context.selectInt(
+            node,
+            QStringLiteral("TrackColorAlpha"),
+            &ok);
+    if (ok) {
```
Comment (Holzhaus) [inline src/widget/wtrackwidgetgroup.cpp:52]: Do we need a range check? 

```suggestion
    if (ok && trackColorAlpha >= 0 && trackColorAlpha <= 255) {
```

**Comment by daschuer** [comment]
AppVeyor timed out. Merge?

File: src/widget/wtrackwidgetgroup.cpp [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+void WTrackWidgetGroup::slotTrackChanged(TrackId trackId) {
+    Q_UNUSED(trackId);
+    updateColor();
+}
+
+void WTrackWidgetGroup::updateColor() {
+    if (m_pCurrentTrack) {
+        m_trackColor = mixxx::RgbColor::toQColor(m_pCurrentTrack->getColor());
+        m_trackColor.setAlpha(m_trackColorAlpha);
```
Comment (uklotzde) [inline src/widget/wtrackwidgetgroup.cpp:87]: Setting the alpha channel on an invalid color does not make any sense. Did you verify that this works as expected and m_trackColor is still invalid then? Please add a debug assertion to check this:

`DEBUG_ASSERT(m_trackColor.isValid() == m_pCurrentTrack->getColor().has_value());`

If has_value() is not supported on all platforms we may either remove it (implicit cast, not sure if this works) or replace it with a `static_cast<bool>`.

File: src/widget/wtrackwidgetgroup.cpp [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        : WWidgetGroup(pParent),
+          m_group(group),
+          m_pConfig(pConfig),
+          m_trackColorAlpha(kDefaultTrackColorAlpha),
+          m_pTrackMenu(make_parented<WTrackMenu>(
+                  this, pConfig, pTrackCollectionManager, kTrackMenuFeatures)) {
+    setAcceptDrops(true);
+}
+
+WTrackWidgetGroup::~WTrackWidgetGroup() {
```
Comment (uklotzde) [inline src/widget/wtrackwidgetgroup.cpp:41]: Please replace with `= default`, either in the .h or the .cpp file depending on the dependencies.

File: src/widget/wtrackwidgetgroup.cpp [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+#include <QStylePainter>
+#include <QUrl>
+
+#include "control/controlobject.h"
+#include "util/dnd.h"
+#include "widget/wtrackmenu.h"
+
+namespace {
+
+int kDefaultTrackColorAlpha = 255;
```
Comment (uklotzde) [inline src/widget/wtrackwidgetgroup.cpp:14]: constexpr

**Review by uklotzde — CHANGES_REQUESTED** [review]
Coding style issues and missing consistency checks

**Comment by Be-ing** [comment]
At first I was sceptical the thin colored line would be too easy to miss. But now that I have tried it, I think this is a good solution that does not stick out too much and does not interfere with the rest of the skin's design.

**Comment by daschuer** [comment]
I think @ronso0 has already ideas. So I don't want to do it here.

**Comment by daschuer** [comment]
The Travis failure is  a MacOs Timout. 
Merge?

File: .travis.yml (hunk 5 above)
Comment (uklotzde) [reply .travis.yml:172]: @daschuer Ping

File: src/widget/wtrackwidgetgroup.cpp (hunk 1 above)
Comment (uklotzde) [reply src/widget/wtrackwidgetgroup.cpp:106]: @daschuer Ping

File: .travis.yml (hunk 5 above)
Comment (daschuer) [reply .travis.yml]: Ups, done.

File: src/widget/wtrackwidgetgroup.cpp (hunk 1 above)
Comment (uklotzde) [reply src/widget/wtrackwidgetgroup.cpp:106]: @daschuer Ping again. You already used testFlag() elsewhere.

Comment (daschuer) [reply src/widget/wtrackwidgetgroup.cpp]: Oh sorry this has slipped through.  Done.

**Comment by Be-ing** [comment]
ping @uklotzde

File: src/widget/wtrackproperty.cpp [hunk 10]
```diff
@@ -78,21 +79,22 @@ void WTrackProperty::updateLabel() {
     setText("");
 }
 
-void WTrackProperty::mouseMoveEvent(QMouseEvent *event) {
-    if ((event->buttons() & Qt::LeftButton) && m_pCurrentTrack) {
+void WTrackProperty::mouseMoveEvent(QMouseEvent* event) {
+    if ((event->buttons().testFlag(Qt::LeftButton)) && m_pCurrentTrack) {
```
Comment (Holzhaus) [inline src/widget/wtrackproperty.cpp:83]:
```suggestion
    if (event->buttons().testFlag(Qt::LeftButton) && m_pCurrentTrack) {
```

File: src/widget/wtracktext.cpp [hunk 11]
```diff
@@ -70,7 +71,7 @@ void WTrackText::updateLabel() {
 }
 
 void WTrackText::mouseMoveEvent(QMouseEvent *event) {
-    if ((event->buttons() & Qt::LeftButton) && m_pCurrentTrack) {
+    if ((event->buttons().testFlag(Qt::LeftButton)) && m_pCurrentTrack) {
```
Comment (Holzhaus) [inline src/widget/wtracktext.cpp:74]:
```suggestion
    if (event->buttons().testFlag(Qt::LeftButton) && m_pCurrentTrack) {
```

File: src/widget/wtrackwidgetgroup.cpp [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+    if (m_trackColor.isValid()) {
+        QStylePainter p(this);
+
+        p.fillRect(rect(), QBrush(m_trackColor));
+    }
+}
+
+void WTrackWidgetGroup::mouseMoveEvent(QMouseEvent* event) {
+    if ((event->buttons().testFlag(Qt::LeftButton)) && m_pCurrentTrack) {
```
Comment (Holzhaus) [inline src/widget/wtrackwidgetgroup.cpp:106]:
```suggestion
    if (event->buttons().testFlag(Qt::LeftButton) && m_pCurrentTrack) {
```

File: src/widget/wtrackwidgetgroup.h [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+  private:
+    void dragEnterEvent(QDragEnterEvent* event) override;
+    void dropEvent(QDropEvent* event) override;
+    void mouseMoveEvent(QMouseEvent* event) override;
+
+    void updateColor();
+
+    const QString m_group;
+    UserSettingsPointer m_pConfig;
```
Comment (Holzhaus) [inline src/widget/wtrackwidgetgroup.h:45]: Can this be const?

**Comment by daschuer** [comment]
Done

**Comment by daschuer** [comment]
@uklotzde @Holzhaus Merge?

**Review by Holzhaus — APPROVED** [review]
LGTM, thank you.

**Comment by Holzhaus** [comment]
@uklotzde Do you want to have another look or should I merge?

**Review by uklotzde — APPROVED** [review]

**Comment by Holzhaus** [comment]
@daschuer Can you pull 2.3 to check if the resolves the pre-commit issues? If so, we can merge.

**Comment by daschuer** [comment]
I can confirmed in works. 
```
[INFO] Checking merge-conflict files only.
```
